### PR TITLE
Functional Tic Tac Toe

### PR DESCRIPTION
**Pull request by SHeimlich**
Because it is Friday and I wanted to start trying out the new features in the API.

### CODE REVIEW

**Comment by piotrswigon** [comment]
Question 1) Should this replace the existing implementation?

**Comment by SHeimlich** [comment]
Probably. Do we want to do that now or once more of the changes to the API are completed?

**Comment by piotrswigon** [comment]
I would say we should be updating in place as features arrive.

I am also not sure we've done the right thing to separate TTT into stages TBH. I would keep just a single version of the TTT and keep it as clean as we can make it with the features that are available.

Currently making API changes require updating the same file up to 6 times, which hurts :(

**Comment by SHeimlich** [comment]
Updated, over 1k lines deleted :open_mouth:

File: particles/Tutorial/Kotlin/Demo/src/TTTGame.arcs [hunk 1]
```diff
@@ -119,7 +119,7 @@ recipe GamePlayersDemoRecipe
     gameState: reads gameState
     myMove: writes playerTwoMove
     player: reads playerTwo
-  description `Kotlin Tutorial TicTacToe One Player Demo`
+  description `Kotlin Functional Tutorial TicTacToe One Player Demo`
```
Comment (csilvestrini) [inline particles/Tutorial/Kotlin/Demo/src/TTTGame.arcs]: You could probably revert these changes if you are just going to replace the existing tutorial.

File: particles/Tutorial/Kotlin/Demo/src/TTTGame.kt [hunk 2]
```diff
@@ -32,6 +32,66 @@ class TTTGame : AbstractTTTGame() {
         arrayOf(2, 4, 6)
     )
 
+    init{
```
Comment (csilvestrini) [inline particles/Tutorial/Kotlin/Demo/src/TTTGame.kt]: `init {`

File: particles/Tutorial/Kotlin/Demo/src/TTTGame.kt [hunk 3]
```diff
@@ -32,6 +32,66 @@ class TTTGame : AbstractTTTGame() {
...
+
+        handles.playerOneMove.onUpdate() { move ->
```
Comment (csilvestrini) [inline particles/Tutorial/Kotlin/Demo/src/TTTGame.kt]: It's more conventional to omit the parens: `onUpdate { move ->`

(Here and throughout file)

File: particles/Tutorial/Kotlin/Demo/src/TTTGame.kt [hunk 4]
```diff
@@ -32,6 +32,66 @@ class TTTGame : AbstractTTTGame() {
...
+        handles.playerTwoMove.onUpdate() { move ->
+            applyMove(
+              move ?: TTTGame_PlayerTwoMove(),
+              handles.playerTwo.fetch() ?: TTTGame_PlayerTwo()
+            )
+        }
+
+        handles.events.onUpdate() {
+            if (hasReset()) {
+                handles.gameState.set(defaultGame)
```
Comment (csilvestrini) [inline particles/Tutorial/Kotlin/Demo/src/TTTGame.kt]: Optional, but probably a nice improvement:

```
with(handles) {
  gameState.set(defaultGame)
  playerOneMove.set(playerOneMove.fetch()!!.copy(move = -1.0))
  playerTwoMove.set(playerTwoMove.fetch()!!.copy(move = -1.0))
  events.clear()
}
```

File: particles/Tutorial/Kotlin/Demo/src/TTTHumanPlayer.kt [hunk 5]
```diff
@@ -14,14 +14,13 @@ package arcs.tutorials.tictactoe
...
 class TTTHumanPlayer : AbstractTTTHumanPlayer() {
-    override fun onHandleUpdate(handle: Handle) {
-        if (handles.events.size <= 0) return
 
-        // Get the element with the largest time as this will be the most recent.
-        val event = handles.events.fetchAll().sortedBy { it.time }.last()
-        // Set the move
-        if (event.type == "move" && event.time > -1.0) {
-            handles.myMove.set(TTTHumanPlayer_MyMove(event.move))
+    init{
```
Comment (csilvestrini) [inline particles/Tutorial/Kotlin/Demo/src/TTTHumanPlayer.kt]: `init {`

File: java/arcs/android/storage/database/DatabaseImpl.kt [hunk 6]
```diff
@@ -425,8 +425,42 @@ class DatabaseImpl(
     override suspend fun delete(
         storageKey: StorageKey,
         originatingClientId: Int?
-    ): Unit = stats.delete.timeSuspending<Unit> {
-        TODO("not implemented")
+    ): Unit = stats.delete.timeSuspending { counters ->
```
Comment (piotrswigon) [inline java/arcs/android/storage/database/DatabaseImpl.kt]: Are these changes supposed to be here? I don't think so...

Comment (SHeimlich) [reply java/arcs/android/storage/database/DatabaseImpl.kt]: No, I pulled and rebased on my laptop and things got funky. Fixed it from my workstation this morning.
